src/extensions/step_handlers/script_handler/src/script_handler.cpp

```diff
@@ -134,25 +139,76 @@ static ADUC_Result Script_Handler_DownloadPrimaryScriptFile(ADUC_WorkflowHandle
     workflowId = workflow_peek_id(handle);
     workFolder = workflow_get_workfolder(handle);
 
-    createResult = ADUC_SystemUtils_MkSandboxDirRecursive(workFolder);
-    if (createResult != 0)
+    // Construct the primary script file path
+    filePath << workFolder << "/" << scriptFileName;
+    scriptFilePath = filePath.str();
+
+    if (!ADUC_HashUtils_GetShaVersionForTypeString(
+            ADUC_HashUtils_GetHashType(entity->Hash, entity->HashCount, 0), &algVersion))
     {
-        Log_Error("Unable to create folder %s, error %d", workFolder, createResult);
-        result = { ADUC_Result_Failure, ADUC_ERC_SCRIPT_HANDLER_CREATE_SANDBOX_FAILURE };
+        Log_Error(
+            "FileEntity for %s has unsupported hash type %s",
+            scriptFilePath.c_str(),
+            ADUC_HashUtils_GetHashType(entity->Hash, entity->HashCount, 0));
+        result.ExtendedResultCode = ADUC_ERC_CONTENT_DOWNLOADER_FILE_HASH_TYPE_NOT_SUPPORTED;
         goto done;
     }
 
-    try
+    // If file exists and has a valid hash, then skip download.
+    // Otherwise, delete an existing file, then download.
+    Log_Debug("Check whether '%s' has already been download into the work folder.", scriptFilePath.c_str());
+
+    if (access(scriptFilePath.c_str(), F_OK) == 0)
```

Comment:
As Nox pointed out offline, seems redundant with same check here: https://github.com/Azure/iot-hub-device-update/blob/main/src/extensions/extension_manager/src/extension_manager.cpp#L883

I think the root issue may be something different that is causing script file to be deleted.